pandas/core/indexes/base.py

```diff
@@ -3696,6 +3696,13 @@ def get_indexer(
         tolerance=None,
     ) -> npt.NDArray[np.intp]:
         method = missing.clean_reindex_fill_method(method)
+        # After _maybe_cast_listlike_indexer, target elements which do not
+        # belong to some category are changed to NaNs
+        # Mask to track actual NaN values compared to inserted NaN values
+        try:
+            target_nans = isna(target)
+        except NotImplementedError:
+            target_nans = False
         target = self._maybe_cast_listlike_indexer(target)
```

Comment:
i think its much more likely that the desired fix will be inside of CategoricalIndex._maybe_cast_listlike_indexer

Reply:
Well, the main changes occur in the file `pandas/core/indexes/category.py:471`. Here, in the function `_maybe_cast_listlike_indexer`, until line number 482 everything seems correct. `codes` variable holds `[1,2,-1]` (indexing the categories 1, 2 and 2 resp). On line 483, the corresponding categories become `[2, 3, NaN]` which seems to be correct behaviour as 4 is not a category and its entry is replaced by a NaN.
When this result is returned to `get_indexer` in the file `pandas/core/indexes/base.py:3691`, that's when problems occur.  Lines 3726 is responsible for a discrepancy in the expected output, when `self` has NaNs. 
When the categorical index is `[1,2,3]` and target is `[2,3,4]`, `self._engine.get_indexer(target.codes)` returns `[1,2,-1]` which is expected.
However, when the categorical index is `[1,2,nan,3]` and target is `[2,3,4]`, `self._engine.get_indexer(target.codes)` returns `[1,3,2]` which is problematic.

Reply:
It seems that the code that had been written here did not account for the possible presence of NaNs in the categorical index. Yet, on line 3728 `loc = self.get_loc(np.nan)`, instead of fixing loc for NaNs to be -1. This indicates that the position of NaN is inferred from the Categorical Index. I assumed there must be some reason for it, and to not disturb that functionality, I wrote my fix around this code.